File: 125.valid-palindrome/solution.go
func isPalindrome(s string) bool {
    var b strings.Builder
    for _, r := range s {
        if unicode.IsLetter(r) || unicode.IsDigit(r) {
            b.WriteRune(unicode.ToLower(r))
        }
    }
    t := b.String()
    i, j := 0, len(t) - 1
    for i < j {
        if t[i] != t[j] {
            return false
        }
        i++
        j--
    }
    return true
}

File: 125.valid-palindrome/solution.kt
class Solution {
    fun isPalindrome(s: String): Boolean {
        val normalized = s.mapNotNull { ch -> when {
            ch.isLetter() -> ch.toLowerCase()
            ch.isDigit() -> ch
            else -> null
        } }.joinToString(separator = "")
        return isPalindromeRec(normalized)
    }
    
    private fun isPalindromeRec(s: String): Boolean =
        when {
            s.length <= 1 -> true
            s.first() != s.last() -> false
            else -> isPalindromeRec(s.substring(1, s.length - 1))
        }
}